src/python/pants/engine/internals/graph.py

```diff
@@ -296,10 +339,43 @@ async def transitive_targets(request: TransitiveTargetsRequest) -> TransitiveTar
         )
 
     return TransitiveTargets(
-        tuple(roots_as_targets), FrozenOrderedSet(visited.difference(transitive_excludes))
+        tuple(dependency_mapping.roots_as_targets),
+        FrozenOrderedSet(dependency_mapping.visited.difference(transitive_excludes)),
     )
 
 
+# -----------------------------------------------------------------------------------------------
+# CoarsenedTargets
+# -----------------------------------------------------------------------------------------------
+
+
+@rule
+async def coarsened_targets(addresses: Addresses) -> CoarsenedTargets:
+    dependency_mapping = await Get(
+        _DependencyMapping,
+        _DependencyMappingRequest(TransitiveTargetsRequest(addresses), expanded_targets=False),
```

Comment:
It'd be helpful to set `include_special_cased_deps=False` so that it's clear you're doing that.

This also makes me wonder if we want `CoarsenedTargetsRequest` so that users can change that behavior. Even if we don't need it now, it may future proof better - I wish we had `TransitiveTargetsRequest` from the start.

Reply:
> It'd be helpful to set `include_special_cased_deps=False` so that it's clear you're doing that.

I'll be honest: I don't know how/why to set that, and it's not documented. So it feels more honest to allow it to default than to set it with a comment that says "I don't know what the right value is". I'm hoping that we can follow up to remove it.

> This also makes me wonder if we want CoarsenedTargetsRequest so that users can change that behavior. Even if we don't need it now, it may future proof better - I wish we had TransitiveTargetsRequest from the start.

Mm. My hope is that we can fix #7490 and remove the need for most `*Request` types instead. But I'm not sure about either future.

Since this API is likely to change a few more times before being stable, I'd rather not do anything too speculative.

Reply:
> I don't know how/why to set that, and it's not documented. So it feels more honest to allow it to default than to set it with a comment that says "I don't know what the right value is". I'm hoping that we can follow up to remove it.

When you have a dep on `relocated_files()`, do you also want to pull in a dep on the original `files()`? For project introspection: yes. For hydrating sources, no.